src/systemd.c

```diff
@@ -0,0 +1,286 @@
+	}
+}
+
+//@todo Improve error messages for create_scope_and_slice & create_scope_user_slice; look at cg_delegate.py
+
+int cgroup_create_scope_and_slice(char *scope_name, char *slice_name, int delegated, char* mode){
+
+	sd_bus_error error = SD_BUS_ERROR_NULL;
+	sd_bus_message *m = NULL, *reply = NULL;
+	char *object = NULL;
+	sd_bus *bus = NULL;
+	int ret = 0;
+	
+	int childPID = fork();
+
+	if(childPID < 0){ //No child process created
+		fprintf(stderr, "fork() failed: %d:%s\n", childPID, strerror(-childPID));
+	}
+	else if (childPID == 0){ //This is the child process, sleep forever
+//		fprintf(stdout, "Child Process PID: %d\n", getpid());
+		permanent_sleep();
+	}
+	else{ //This is the parent process, move child process to the desired slice/scope
+
+//		fprintf(stdout, "Parent Process PID: %d\n", getpid());
+		
+		ret = sd_bus_open_system(&bus);
+		if (ret<0){
+			fprintf(stderr, "sd_bus_open_system failed: %d:%s\n", ret, strerror(-ret));
+			goto out;
+		}
+
+		ret = sd_bus_message_new_method_call(bus, &m, "org.freedesktop.systemd1",
+			"/org/freedesktop/systemd1", "org.freedesktop.systemd1.Manager",
+			 "StartTransientUnit");
+		if (ret<0){
+			fprintf(stderr, "sd_bus_message_new_method_call failed: %d:%s\n", ret, strerror(-ret));
+			goto out;
+		}
+
+		ret = sd_bus_message_append(m, "ss", scope_name, mode);
+		if (ret<0){
+			fprintf(stderr, "1st sd_bus_message_append failed: %d:%s\n", ret, strerror(-ret));
+			goto out;
+		}
+
+		ret = sd_bus_message_open_container(m, 'a', "(sv)");
+		if (ret<0)
+			{fprintf(stderr, "sd_bus_message_open_container failed: %d:%s\n", ret, strerror(-ret));
+			goto out;
+		}
+
+		ret = sd_bus_message_append(m, "(sv)", "PIDs", "au", 1, (uint32_t) childPID);
+		if (ret<0){
+			fprintf(stderr, "2nd sd_bus_message_append failed: %d:%s\n", ret, strerror(-ret));
+			goto out;
+		}
+
+		if(delegated == 1){
+			ret = sd_bus_message_append(m, "(sv)", "Delegate", "b", 1);
+			if (ret<0){
+				fprintf(stderr, "3rd sd_bus_message_append failed: %d:%s\n", ret, strerror(-ret));
+				goto out;
+			}
+		}
+
+		ret = sd_bus_message_append(m, "(sv)", "Slice", "s", slice_name);
+		if (ret<0){
+			fprintf(stderr, "4th sd_bus_message_append failed: %d:%s\n", ret, strerror(-ret));
+			goto out;
+		}
+
+		ret = sd_bus_message_close_container(m);
+		if (ret<0){
+			fprintf(stderr, "sd_bus_message_close failed: %d:%s\n", ret, strerror(-ret));
+			goto out;
+		}
+
+		ret = sd_bus_message_append(m, "a(sa(sv))", 0);
+		if (ret<0){
+			fprintf(stderr, "5th sd_bus_message_append failed: %d:%s\n", ret, strerror(-ret));
+			goto out;
+		}
+
+		ret = sd_bus_call(bus, m, 0, &error, &reply);
+		if (ret<0){
+			fprintf(stderr, "sd_bus_call failed: %d:%s\n", ret, strerror(-ret));
+			goto out;
+		}
+
+		ret = sd_bus_message_read(reply, "o", &object);
+		if (ret<0){
+			fprintf(stderr, "sd_bus_message_reply failed: %d:%s\n", ret, strerror(-ret));
+			goto out;
+		}
+		
+		// Parent thread will return PID of child thread
+		ret = childPID;
```

Comment:
The caller should be told the PID of the permanent sleeper process (or else given some way to find it), else the caller has no clue which process to kill (and may kill itself--child and parent have the same name!). While it isn't necessary to save the user the mishap, isn't it safer to do so?

Reply:
Yes, sharing the PID with the caller would be nice, but overloading a return call is dangerous.  It would probably be better to have it as an out parameter in the function call.  Something like this:

```
 int cgroup_create_scope_and_slice(char *scope_name, char *slice_name, int delegated, char* mode, pid_t * const placeholder_pid){ 
 ```
 
(The const is optional, but I'm a big fan of `const` :)

Reply:
Fixed, with const